tests/test_regex.py

```diff
@@ -1,37 +1,267 @@
+            unicodedata.category(c) in UNICODE_WORD_CATEGORIES
+        )
+        for c in s
+    )
+
+
+class TestRegexUnicodeMatching:
+    def _test_matching_pattern(self, pattern, isvalidchar, unicode=False):
+        if unicode:
+            r = re.compile(pattern) if six.PY3 else re.compile(pattern, re.UNICODE)
+        else:
+            r = re.compile(pattern, re.ASCII) if six.PY3 else re.compile(pattern)
+
+        codepoints = six.moves.range(0, sys.maxunicode+1) \
+            if unicode else six.moves.range(1, 128)
+        for c in [six.unichr(x) for x in codepoints]:
+            if isvalidchar(c):
+                assert r.match(c), (
+                    '"%s" supposed to match "%s" (%r, category "%s"), '
+                    'but it doesnt' % (pattern, c, c, unicodedata.category(c))
+                )
+            else:
+                assert not r.match(c), (
+                    '"%s" supposed not to match "%s" (%r, category "%s"), '
+                    'but it does' % (pattern, c, c, unicodedata.category(c))
+                )
+
+    def test_matching_ascii_word_chars(self):
+        self._test_matching_pattern('\\w', is_word)
+
+    def test_matching_unicode_word_chars(self):
+        self._test_matching_pattern('\\w', is_word, unicode=True)
+
+    def test_matching_ascii_non_word_chars(self):
+        self._test_matching_pattern('\\W', lambda s: not is_word(s))
+
+    def test_matching_unicode_non_word_chars(self):
+        self._test_matching_pattern('\\W', lambda s: not is_word(s), unicode=True)
+
+    def test_matching_ascii_digits(self):
+        self._test_matching_pattern('\\d', is_digit)
+
+    def test_matching_unicode_digits(self):
+        self._test_matching_pattern('\\d', is_digit, unicode=True)
+
+    def test_matching_ascii_non_digits(self):
+        self._test_matching_pattern('\\D', lambda s: not is_digit(s))
+
+    def test_matching_unicode_non_digits(self):
+        self._test_matching_pattern('\\D', lambda s: not is_digit(s), unicode=True)
+
+    def test_matching_ascii_spaces(self):
+        self._test_matching_pattern('\\s', is_space)
+
+    def test_matching_unicode_spaces(self):
+        self._test_matching_pattern('\\s', is_unicode_space, unicode=True)
+
+    def test_matching_ascii_non_spaces(self):
+        self._test_matching_pattern('\\S', lambda s: not is_space(s))
+
+    def test_matching_unicode_non_spaces(self):
+        self._test_matching_pattern('\\S', lambda s: not is_unicode_space(s), unicode=True)
 
 
 def assert_can_generate(pattern):
-    strategy = hypothesis_regex.regex(pattern)
-    for _ in six.moves.range(100):
-        s = strategy.example()
-        assert re.match(pattern, s) is not None, \
-            '"%s" supposed to match "%s" (strategy = %s)' % (s, pattern, strategy)
+    compiled_pattern = re.compile(pattern)
+    strategy = regex(pattern)
+
+    h.find(strategy, lambda s: compiled_pattern.match(s) is not None)
+
+    with pytest.raises(he.NoSuchExample):
```

Comment:
Well, this is the main purpose of the assert: check that every generated string built from given regex matches regex. I will actually remove the previous positive case (the find() above) to make it more clear, I think that find() is redundant.